### ISSUE TITLE
how to derive candidates for two sources from one single asynchronous process?

### ISSUE THREAD

**Issue by dfeich** [emacs-helm/helm#1781]
Dear Thierry

How would you solve the problem of having two sources where both depend on candidates produced in the same asynchronous process?

In the async process I am doing a recursive grep for finding all files that contain a list of search expressions. Now I want to display the filenames of the hits in one source and the full list of lines that matched in the second source. So, somehow I would need one source to wait for another source.

In the past I was able to cheat my way around it by abusing the :delay switch for the source which should use results from the other source (terrible hack, I agree, and I think it makes sense you removed delay). But could you give me a hint how one could synchronize between the two async sources in some way?

My current code resides in https://github.com/dfeich/helm-deft.

## Describe versions of helm, emacs, operating system etc.
I am using helm-core-20170514.2238 and helm-20170513.2252 from MELPA

Thanks for a hint,
Derek

**Comment by thierryvolpiatto** [comment]
Derek <notifications@github.com> writes:

> In the async process I am doing a recursive grep for finding all files
> that contain a list of search expressions. Now I want to display the
> filenames of the hits in one source and the full list of lines that
> matched in the second source. So, somehow I would need one source to
> wait for another source.

What about feeding a variable with all the filenames fetched from the
grep buffer process in the sentinel of this process ?

e.g

(when (string= event "finished\n")
  (with-current-buffer (process-buffer process)
     (setq my-var-to-feed-candidates-in-second-source
           <code to collect fnames>)))

(defvar my-second-source ([....] 
                          :candidates
                          'my-var-to-feed-candidates-in-second-source
                          [....]))

The variable would then be used in the :candidates slot for the second
source.  Perhaps reset the var in the init fn of first source.

Fully not tested, just a thought.

--
Thierry
Gpg Key fingerprint = 6CEC 7081 AB33 E251 4AB8  5FC2 28D1 7F53 59F2 9997

**Comment by thierryvolpiatto** [comment]
However, I am not sure this is the right approach to achieve what you want as each time you search for a file the grep process will kick in again needlessly.

**Comment by dfeich** [comment]
Thanks for the fast answer

I think I am already doing something that is quite similar to your
suggestion. I am (ab)using the :filter-one-by-one setting of the source
that is  doing the asynchronous grep to build up the (global) candidates
list  that is used in the second source. I can see that the filter function
is called via helm--maybe-process-filter-one-by-one-candidate in the output
filter to the asynchronous process of the first source.

I guess that my main problem is that the async source is finished last and
the other source which should use the side product of the async source
probably already used the still empty candidate list. In the end, I would
like to be able and trigger an update of just that second source at the
moment where the async source is finished.

I played a bit with a sentinel (I already had one from my earlier
implementation 2y ago) and thereby learned that I also need to cover as
well the "killed" outcome that is triggered when the source has too many
entries. Helm is quite complex, but it is nice training to understand
process handling.


On Sun, Jun 4, 2017 at 12:26 PM, Thierry Volpiatto <notifications@github.com
> wrote:

>

**Comment by dfeich** [comment]
I think your tip is good (q.v. my other answer), but I am already doing something like it.

In regard to the needles searching, maybe I did not explain enough what the code is doing: The user defines in the configuration some directories and a file extension (for me typically I define *.org to have the grep only search my org files).

The user can type in the helm line a number of search words, and I am building up a recursive bash expression of nested greps where each outer grep is running on the files which were identified for the inner grep. The outermost grep then is the only one which returns a list of lines with matches. I have a function building up that command which takes the search string and the list of files to search.

    (helm-deft-build-cmd "one two three" '("file1" "file2" "file3"))

And this would be the generated shell command from this example

    "grep  -EHine 'three' $(grep  -Elie 'two' $(grep  -Elie 'one' file1 file2 file3))"

When this command has run in the async source, I display the grepped lines in that source. And I can compose from this the list of files that matched which I want to display in a second source. But the update of that second source should happen after the async process for the first source has finished to also compose the candidates list for that second source.

Thanks.

**Comment by thierryvolpiatto** [comment]
Derek <notifications@github.com> writes:

> Thanks for the fast answer
>
> I think I am already doing something that is quite similar to your
> suggestion. I am (ab)using the :filter-one-by-one setting of the source
> that is doing the asynchronous grep to build up the (global) candidates
> list that is used in the second source.

This is not similar, your filter will run after each chunk of output is
coming and leave you only the last wave of output.

As example (even if unrelated to your code) see helm-redisplay-buffer
and how it is used in helm-locate to sort candidate when process is
finished.


> I can see that the filter function is called via
> helm--maybe-process-filter-one-by-one-candidate in the output filter
> to the asynchronous process of the first source.
>
> I guess that my main problem is that the async source is finished last and
> the other source which should use the side product of the async source
> probably already used the still empty candidate list. In the end, I would
> like to be able and trigger an update of just that second source at the
> moment where the async source is finished.

Probably you should then change your approach and use synchronous source ?
Anyway you have to wait the first finish isn't it ?
So why using an async source in first place ?


-- 
Thierry
Gpg Key fingerprint = 6CEC 7081 AB33 E251 4AB8  5FC2 28D1 7F53 59F2 9997

**Comment by dfeich** [comment]
You are completely  correct. I had introduced the async process originally when I thought that the grep may take a long time and that the results would come heapwise. But this is not true in the way I do it now. So I certainly could use a sync source.

But I also was able now to resolve the problem while even using the async source. I needed to hand the variable in quoted form to the :candidates slot of the sync source. Somehow I did not transform this correctly when I did the adaptions for the new object oriented style. The code works now fine, even though I probably should use sync sources.

Thanks for your help and for providing this highly useful package!
Derek
